File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.java
// Last updated: 14/08/2025, 11:35:43
/*class Solution {
    public int lengthOfLongestSubstring(String s) {
        
    }
}*/
import java.util.HashSet;
import java.util.Set;

public class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int maxLength = 0;
        int left = 0;

        for (int right = 0; right < s.length(); right++) {
            char currentChar = s.charAt(right);

            // If character is in the set, move left pointer to remove duplicates
            while (set.contains(currentChar)) {
                set.remove(s.charAt(left));
                left++;
            }

            // Add current character to set and update max length
            set.add(currentChar);
            maxLength = Math.max(maxLength, right - left + 1);
        }

        return maxLength;
    }
}

File: 3-LongestSubstringWithoutRepeatingCharacters/3-LongestSubstringWithoutRepeatingCharacters.py
# Last updated: 09/09/2026, 16:38:20
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set=set()
        l=0
        maxlen=0

        for r in range(len(s)):
            while s[r] in char_set:
                char_set.remove(s[l])
                l+=1
            char_set.add(s[r])
            maxlen=max(maxlen,r-l+1)
        return maxlen    